aiida_quantumespresso/utils/validation/trajectory.py

```diff
@@ -54,14 +74,24 @@ def verify_convergence_forces(trajectory, index=-1, threshold=None):
     threshold *= CONSTANTS.ry_to_ev / CONSTANTS.bohr_to_ang  # Convert to eV / Å
 
     try:
-        forces = trajectory.get_array('forces')[index]
+        abs_forces = numpy.abs(trajectory.get_array('forces')[index])
     except (KeyError, IndexError) as exception:
         raise ValueError('the `forces` array does not exist or the given index exceeds the length.') from exception
 
-    return numpy.max(abs(forces)) < threshold
+    if fixed_coords is not None:
+        fixed_coords = numpy.array(fixed_coords)
+        # Set the forces corresponding to fixed coordinates to zero, as they should not be checked versus threshold
+        abs_forces *= numpy.invert(fixed_coords)
```

Comment:
Deducing from what you are doing here, I take it the `fixed_coords` is a list of lists where each internal member is a list of three elements with each element either a 0 or a 1? If that is the case, I think it would be good to have a short explanation of this and we should add validation of `fixed_coords`. I am not sure how liberal the format is in the actual `settings` of the `PwCalculation`. For example, maybe QE accepts any positive non-zero integer to fix a coordinate, and so if a user uses `2` instead of `1`, would `invert` still work as intended?

Reply:
Actually the `fixed_coords` setting is a list of lists of booleans, e.g.:

```python
settings_dict = {
    'fixed_coords': [
        [True, False, False],
        [True, True, True],
        [False, False, False],
        [False, False, False],
        [False, False, False]
    ],
}
```

See also [this section](https://aiida-quantumespresso.readthedocs.io/en/latest/user_guide/calculation_plugins/pw.html?highlight=fixed_coords#fixing-some-atom-coordinates) in the documentation. The fact that these must be booleans is enforced during the `BasePwCpInputGenerator._generate_PWCPinputdata` step:

https://github.com/aiidateam/aiida-quantumespresso/blob/0c8a6114dfb740262fd408c7b0fc73bd0fde33ee/aiida_quantumespresso/calculations/__init__.py#L492-L497

However, this would of course result in an `Excepted` work chain, so I agree adding a validator of the `settings` input is an improvement. Will adapt (and also clean up this code a bit). 👍 

Will also extend the explanation and improve the type hinting for clarity.